File: src/main/java/com/cxxcxx/zinecraft/core/structure/ModStructure.java
package com.cxxcxx.zinecraft.core.structure;

import com.cxxcxx.zinecraft.api.world.structure.JigsawBuildingEntry;
import com.cxxcxx.zinecraft.core.Zinecraft;
import net.minecraft.world.level.biome.Biomes;
import net.minecraft.world.level.levelgen.GenerationStep.Decoration;
import net.minecraft.world.level.levelgen.Heightmap.Types;
import net.minecraft.world.level.levelgen.structure.TerrainAdjustment;
import net.minecraft.world.level.levelgen.structure.pools.StructureTemplatePool.Projection;

public final class ModStructure {
  public static final JigsawBuildingEntry STARGATE = Zinecraft.WORLDGEN.structures.jigsawBuilding(
      "stargate",
      "雪原星门",
      "Snowfield Stargate",
      8,
      4,
      958853901,
      1,
      32,
      0.0F,
      Biomes.SNOWY_PLAINS,
      false,
      0,
      Types.WORLD_SURFACE_WG,
      0,
      false,
      false,
      Decoration.SURFACE_STRUCTURES,
      TerrainAdjustment.BEARD_THIN,
      builder -> builder.pool("start", Projection.RIGID, pool -> pool.template("stargate", 1))
  );

  private ModStructure() {
  }

  public static void bootstrap() {
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/core/structure/ModStructure.kt
package com.cxxcxx.zinecraft.core.structure

import com.cxxcxx.zinecraft.core.Zinecraft

object ModStructure {
  val PORTAL_RUINS_COMMON = Zinecraft.STRUCTURES.simpleBuilding(
    path = "portal_ruins_common",
    template = "portal_ruins/common",
    spacing = 36,
    separation = 30,
    salt = 958853901,
    removeVinesChance = 0.6f
  )

  /** start -> middle -> end 三个 Jigsaw 结构片段。 */
  val THREE_PIECE_JIGSAW = Zinecraft.STRUCTURES.jigsawBuilding(
    path = "jigsaw_example",
    spacing = 40,
    separation = 20,
    salt = 31579842,
    size = 2
  ) {
    pool("start") {
      template("jigsaw_example/start")
    }
    pool("middle") {
      template("jigsaw_example/middle")
    }
    pool("end") {
      template("jigsaw_example/end")
    }
  }
}
